### PR TITLE
MCA/UCX: added common module

### PR DESCRIPTION
**Pull request by hoopoepg**
- implemented non-blocking routines for flush operations

Signed-off-by: Sergey Oblomov <sergeyo@mellanox.com>

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e5c32a31edc4ec1c3bd2adb692be6cb9

**Comment by hoopoepg** [comment]
@yosefe here is implementation of common functionality for UCX (part of it).
what do you think?

**Comment by yosefe** [comment]
@jladd-mlnx @xinzhao3 @alinask FYI

File: opal/mca/common/ucx/common_ucx.c [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "common_ucx.h"
+
+/***********************************************************************/
+
+static void opal_common_ucp_send_cb(void *request, ucs_status_t status)
+{
+}
+
+ucs_status_t opal_common_ucx_ep_flush(ucp_ep_h ep, ucp_worker_h worker)
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply opal/mca/common/ucx/common_ucx.c]: done

File: opal/mca/common/ucx/common_ucx.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+OPAL_DECLSPEC ucs_status_t opal_common_ucx_worker_flush(ucp_worker_h worker);
+
+static inline
+ucs_status_t opal_common_ucx_wait_request(ucs_status_ptr_t request, ucp_worker_h worker)
+{
+    ucs_status_t status;
+    int i;
+
+    /* check for request completed or failed */
+    if (UCS_OK == request) {
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply opal/mca/common/ucx/common_ucx.h]: fixed

File: opal/mca/common/ucx/common_ucx.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        return UCS_OK;
+    } else if (UCS_PTR_IS_ERR(request)) {
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply opal/mca/common/ucx/common_ucx.h]: fixed

File: opal/mca/common/ucx/common_ucx.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        return UCS_PTR_STATUS(request);
+    }
+
+    while (1) {
+        /* call UCX progress */
+        for (i = 0; i < 100; i++) {
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply opal/mca/common/ucx/common_ucx.h]: added parameter opal_common_ucx_progress_iterations to setup iterations number

File: opal/mca/common/ucx/common_ucx.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            if (UCS_INPROGRESS != (status = ucp_request_check_status(request))) {
+                ucp_request_free(request);
+                return status;
+            }
+            ucp_worker_progress(worker);
+        }
+        /* call OPAL progress on every 100 call to UCX progress */
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply opal/mca/common/ucx/common_ucx.h]: fixed

File: oshmem/mca/atomic/ucx/atomic_ucx.h [hunk 6]
```diff
@@ -66,28 +67,7 @@ void mca_atomic_ucx_complete_cb(void *request, ucs_status_t status);
...
-            if (UCS_INPROGRESS != (status = ucp_request_check_status(request))) {
-                ucp_request_free(request);
-                return status;
-            }
-            ucp_worker_progress(mca_spml_self->ucp_worker);
-        }
-        /* call OPAL progress on every 100 call to UCX progress */
-        opal_progress();
-    }
+    return opal_common_ucx_wait_request(request, mca_spml_self->ucp_worker);
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply oshmem/mca/atomic/ucx/atomic_ucx.h]: removed

File: opal/mca/common/ucx/common_ucx.c [hunk 7]
```diff
@@ -10,25 +10,29 @@
...
+    static int registered = 0;
+
+    if (registered) {
+        /* process once */
+        return;
+    }
+
+    registered = 1;
+    mca_base_var_register("opal", "opal_common", "ucx", "progress_iterations",
+            "Set number of calls of internal UCX progress calls per opal_progress call",
```
Comment (yosefe) [inline opal/mca/common/ucx/common_ucx.c]: minor: indentation

File: opal/mca/common/ucx/common_ucx.h [hunk 8]
```diff
@@ -73,6 +73,44 @@ ucs_status_t opal_common_ucx_worker_flush(ucp_worker_h worker)
...
+
+    status = opal_common_ucx_atomic_fetch(ep, UCP_ATOMIC_FETCH_OP_CSWAP, compare, &tmp,
+                                          op_size, remote_addr, rkey, worker);
+    if (OPAL_LIKELY(UCS_OK == status)) {
+        /* in case if op_size is constant (like sizeof(type)) then this condition
+         * is evaluated in compile time */
+        if (op_size == sizeof(uint64_t)) {
+            *(uint64_t*)result = tmp;
+        } else {
+            assert(op_size == sizeof(uint32_t));
```
Comment (yosefe) [inline opal/mca/common/ucx/common_ucx.h]: is there something like opal_assert() ?

Comment (hoopoepg) [reply opal/mca/common/ucx/common_ucx.h]: can't find any global opal-specific assert... most of the components use standard assert command

**Comment by yosefe** [comment]
bot:retest

**Comment by hoopoepg** [comment]
@yosefe ok to go?
it prevents atomic refactoring

**Comment by hoopoepg** [comment]
👍

**Comment by hoopoepg** [comment]
@ggouaillardet sorry for that.
I added PR to fix issue: https://github.com/open-mpi/ompi/pull/5344
could you try it to build OMPI in your environment?

thank you

**Pull request merged**